core/src/main/java/me/aurium/beetle/api/task/sync/SyncFuture.java

```diff
@@ -0,0 +1,72 @@
+package me.aurium.beetle.api.task.sync;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.function.Supplier;
+
+/**
+ * i want to stop existing
+ *
+ * A safer to use Future that permits get and join based on an internal behavior.
+ *
+ * TODO eventually some sort of resync to main thread functionality
+ *
+ * @param <T>
+ */
+public class SyncFuture<T> extends CompletableFuture<T> implements Synced {
+
+    private final Thread mainThread;
+    private final SyncQueue activeQueue;
+    private final BlockingBehavior behavior;
+
+    public SyncFuture(Thread mainThread, SyncQueue activeQueue, BlockingBehavior behavior) {
+        this.mainThread = mainThread;
+        this.activeQueue = activeQueue;
+        this.behavior = behavior;
+    }
+
+    //TODO a248 i don't know if this is necessary but given that get operates on a similar principle (wait on the main thread if we are running sync)
+    //i think it might be required. Anyways review this and see if it is necessary
+    @Override
+    public T get() throws ExecutionException, InterruptedException {
+        if (isOnMainThread()) {
+            while (!isDone()) {
+                activeQueue.pulseQueue();
+
+                if (isDone()) {
+                    return super.get();
+                }
+
+                behavior.behavior();
```

Comment:
Unfortunately, it is not so simple as this to design a BlockingBehavior API. Your BlockingBehavior really only permits one implementation, and that is the implementation you currently have, TimerSleepBehavior.

You're forbidding a BlockingBehavior implementation which uses a Lock/Condition or synchronized/wait/notify. These kinds of implementations are the most straightforward and non-controversial ones. They cannot go wrong and are often used as a baseline for comparison to non-blocking approaches. Any reasonable API would accommodate them.

To see why you're preventing a BlockingBehavior based on wait/notify semantics, observe two things: first, you need a way for the BlockingBehavior to be notified. Second, the BlockingBehavior needs to have more extensive control of the wait loop, in order to prevent a race condition between checking the completion status of the future and beginning to wait on the future's completion.